rfcs/0003-nested-field-selection.md

```diff
@@ -0,0 +1,61 @@
+# Nested Field Selection
+
+## Purpose
+
+Field selection is useful in a variety of contexts:
+
+- Document databases contain nested structures, and more often than not, we only want to select a subset of the available data.
+- Many relational databases also have semi-structured datatypes like `json(b)`.
+- The TypeScript connector can return object and array types, and we can pass along the selection set and only return the requested data.
+
+This proposal is a non-breaking, minimal addition to the specification, which adds _selection_ for nested structures. Aadditional operations on this sort of data (filtering, sorting, aggregation, etc.) will be covered in later proposals.
+
+## Proposal 
+
+We extend `Field::Column` with an optional `NestedField` property which allows for selection within nested arrays and fields of nested objects:
+
+```rust
+pub enum Field {
+    Column {
+        column: String,
+        nested_field: Option<NestedField>
+    },
+    Relationship {
+        query: Box<Query>,
+        /// The name of the relationship to follow for the subquery
+        relationship: String,
+        /// Values to be provided to any collection arguments
+        arguments: BTreeMap<String, RelationshipArgument>,
+    },
+}
+```
+
+If `nested_field == None` then the column is returned in its entirety (ensuring backwards-compatibility).
+
+A `NestedField` can be either an object or an array:
+
+```rust
+pub enum NestedField {
+    Object(NestedObject),
+    Array(NestedArray)
+}
+
+pub struct NestedObject {
+    pub fields: IndexMap<String, Field>,
+}
+
+pub struct NestedArray {
+    pub field: Box<NestedField>,
```

Comment:
@dmoverton Note, I omitted the `Option` here, until we figure that detail out.

Reply:
To clarify my previous comment: at the moment `Column {column: "name", fields: Some(Array(NestedArray{ fields: None }))}` is equivalent to `Column {column: "name", fields: None }` so could argue that there's no point in making `NestedArray.field` optional. However, in future we probably want to extend `NestedArray` with `limit`, `offset`, `order_by` and `predicate`:
```rust
pub struct NestedArray {
    pub fields: Box<Option<NestedField>>,
    pub limit: Option<u32>,
    /// Optionally offset from the Nth result
    pub offset: Option<u32>,
    pub order_by: Option<OrderBy>,
    #[serde(rename = "where")]
    pub predicate: Option<Expression>,
}
```
(We have already done this is v2: see https://gist.github.com/dmoverton/924c069dded11c2101db54e8afd32b26)
If/when we do this we will want `NestedArray.fields` to be optional so we can have those other things without requiring an element subselection. So I guess the question is what we want to do right now? I'm happy to change it to be non-optional if you like, but I also don't have a problem with leaving it how it is.

Reply:
Ah, got it. That makes sense. Would you mind updating the note in the RFC to this effect then?

Reply:
Do we need to rename it to `fields` then?

Reply:
Sorry, thinking about this some more, doesn't this mean the right solution is still to remove the outer `Option`? Specifically, there is still duplication between

```rust
Column { column: "name", fields: Some(Array(NestedArray{ fields: None, limit: None, offset: None, order_by: None, predicate: None })) }
```

and

```rust
Column { column: "name", fields: None }
```

Both return the whole array. We can't require all of the inner `Option`s to be not-simultaneously-null, but we can remove the outer `Option`.

Reply:
Technically, we could remove the outer `Option`, but then all scalar column selections would need to be
```rust
Column { column: "name", fields: Array(...) }
```
which seems a bit odd, unless we can come up with a better name for the `Array` enum. It possibly also presents some backwards compatibility issues, if we have a connector using a newer version of the spec than the engine (although maybe we don't care about that).

Reply:
Ok, so ignore the comment about removing the outer `Option`, I think I was confusing myself about something.

It will be non-breaking to add the `Option` later when we implement aggregates, but at that point we will need to figure out the return type for nested aggregates anyway. So we could remove the `Option` now with a plan to add it back in a non-breaking change later if/when we figure out that return type. Or we could assume we'll figure it out and just keep it now. I'd probably prefer the first option to minimize confusion about what the extra `Option` is for.

Reply:
@paf31 I've made the change you suggested to remove the `Option` from `NestedArray`.